File: allure2-androidj/src/main/java/com/yq/allure2_androidj/common/AllureLifecycle.java
package com.yq.allure2_androidj.common;

import android.util.Log;

import com.yq.allure2_androidj.android.listener.LifecycleNotifier;
import com.yq.allure2_androidj.common.lowJdk.Consumer;
import com.yq.allure2_androidj.common.lowJdk.Optional;
import com.yq.allure2_androidj.common.resultRW.AllureResultsWriter;
import com.yq.allure2_androidj.common.resultRW.FileAndroidResultsWriter;
import com.yq.allure2_androidj.common.utils.Objects;
import com.yq.allure2_androidj.model.Attachment;
import com.yq.allure2_androidj.model.FixtureResult;
import com.yq.allure2_androidj.model.Stage;
import com.yq.allure2_androidj.model.StepResult;
import com.yq.allure2_androidj.model.TestResult;
import com.yq.allure2_androidj.model.TestResultContainer;
import com.yq.allure2_androidj.model.WithAttachments;
import com.yq.allure2_androidj.model.listeners.ContainerLifecycleListener;
import com.yq.allure2_androidj.model.listeners.FixtureLifecycleListener;
import com.yq.allure2_androidj.model.listeners.StepLifecycleListener;
import com.yq.allure2_androidj.model.listeners.TestLifecycleListener;


import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.util.List;
import java.util.UUID;

import static com.yq.allure2_androidj.common.resultRW.AllureConstants.ATTACHMENT_FILE_SUFFIX;
import static com.yq.allure2_androidj.common.resultRW.AllureConstants.TAG;
import static com.yq.allure2_androidj.common.utils.ServiceLoaderUtils.load;
import static com.yq.allure2_androidj.common.utils.Tools.isNone;
import static java.sql.DriverManager.println;


/**
 * The class contains Allure context and methods to change it.
 */
@SuppressWarnings("PMD.TooManyMethods")
public class AllureLifecycle {

    private final AllureResultsWriter writer;

    private final AllureStorage storage;

    private final LifecycleNotifier notifier;

    public AllureLifecycle() {
        this(new FileAndroidResultsWriter());
    }

    public AllureLifecycle(final AllureResultsWriter writer) {
        final ClassLoader classLoader = getClass().getClassLoader();
        this.notifier = new LifecycleNotifier(
                load(ContainerLifecycleListener.class, classLoader),
                load(TestLifecycleListener.class, classLoader),
                load(FixtureLifecycleListener.class, classLoader),
                load(StepLifecycleListener.class, classLoader)
        );
        this.writer = writer;
        this.storage = new AllureStorage();
    }

    public void startTestContainer(final String parentUuid, final TestResultContainer container) {
        updateTestContainer(parentUuid, found -> found.getChildren().add(container.getUuid()));
        startTestContainer(container);
    }

    public void startTestContainer(final TestResultContainer container) {
        notifier.beforeContainerStart(container);
        container.setStart(System.currentTimeMillis());
        storage.addContainer(container);
        notifier.afterContainerStart(container);
    }

    public void updateTestContainer(final String uuid, final Consumer<TestResultContainer> update) {
        storage.getContainer(uuid).ifPresent(container -> {
            notifier.beforeContainerUpdate(container);
            update.accept(container);
            notifier.afterContainerUpdate(container);
        });
    }

    public void stopTestContainer(final String uuid) {
        storage.getContainer(uuid).ifPresent(container -> {
            notifier.beforeContainerStop(container);
            container.setStop(System.currentTimeMillis());
            notifier.afterContainerUpdate(container);
        });
    }

    public void writeTestContainer(final String uuid) {
        storage.removeContainer(uuid).ifPresent(container -> {
            notifier.beforeContainerWrite(container);
            writer.write(container);
            notifier.afterContainerWrite(container);
        });
    }

    public void startPrepareFixture(final String parentUuid, final String uuid, final FixtureResult result) {
        notifier.beforeFixtureStart(result);
        updateTestContainer(parentUuid, container -> container.getBefores().add(result));
        startFixture(uuid, result);
        notifier.afterFixtureStart(result);
    }

    public void startTearDownFixture(final String parentUuid, final String uuid, final FixtureResult result) {
        notifier.beforeFixtureStart(result);
        updateTestContainer(parentUuid, container -> container.getAfters().add(result));
        startFixture(uuid, result);
        notifier.afterFixtureStart(result);
    }

    private void startFixture(final String uuid, final FixtureResult result) {
        storage.addFixture(uuid, result);
        result.setStage(Stage.RUNNING);
        result.setStart(System.currentTimeMillis());
        storage.clearStepContext();
        storage.startStep(uuid);
    }

    public void updateFixture(final Consumer<FixtureResult> update) {
        updateFixture(storage.getRootStep(), update);
    }

    public void updateFixture(final String uuid, final Consumer<FixtureResult> update) {
        storage.getFixture(uuid).ifPresent(fixture -> {
            notifier.beforeFixtureUpdate(fixture);
            update.accept(fixture);
            notifier.afterFixtureUpdate(fixture);
        });
    }

    public void stopFixture(final String uuid) {
        storage.removeFixture(uuid).ifPresent(fixture -> {
            notifier.beforeFixtureStop(fixture);
            storage.clearStepContext();
            fixture.setStage(Stage.FINISHED);
            fixture.setStop(System.currentTimeMillis());
            notifier.afterFixtureStop(fixture);
        });
    }

    public Optional<String> getCurrentTestCase() {
        return Optional.ofNullable(storage.getRootStep());
    }

    public void scheduleTestCase(final String parentUuid, final TestResult result) {
        updateTestContainer(parentUuid, container -> container.getChildren().add(result.getUuid()));
        scheduleTestCase(result);
    }

    public void scheduleTestCase(final TestResult result) {
        notifier.beforeTestSchedule(result);
        result.setStage(Stage.SCHEDULED);
        storage.addTestResult(result);
        notifier.afterTestSchedule(result);
    }

    public void startTestCase(final String uuid) {
        storage.getTestResult(uuid).ifPresent(testResult -> {
            notifier.beforeTestStart(testResult);
            testResult
                    .withStage(Stage.RUNNING)
                    .withStart(System.currentTimeMillis());
            storage.clearStepContext();
            storage.startStep(uuid);
            notifier.afterTestStart(testResult);
        });
    }

    public void updateTestCase(final Consumer<TestResult> update) {
        final String uuid = storage.getRootStep();
        updateTestCase(uuid, update);
    }

    public void updateTestCase(final String uuid, final Consumer<TestResult> update) {
        storage.getTestResult(uuid).ifPresent(testResult -> {
            notifier.beforeTestUpdate(testResult);
            update.accept(testResult);
            notifier.afterTestUpdate(testResult);
        });
    }

    public void stopTestCase(final String uuid) {
        storage.getTestResult(uuid).ifPresent(testResult -> {
            notifier.beforeTestStop(testResult);
            testResult
                    .withStage(Stage.FINISHED)
                    .withStop(System.currentTimeMillis());
            storage.clearStepContext();
            notifier.afterTestStop(testResult);
        });
    }

    public void writeTestCase(final String uuid) {
        storage.removeTestResult(uuid).ifPresent(testResult -> {
            notifier.beforeTestWrite(testResult);
            writer.write(testResult);
            notifier.afterTestWrite(testResult);
        });
    }

    public void startStep(final String uuid, final StepResult result) {
        storage.getCurrentStep().ifPresent(parentUuid -> startStep(parentUuid, uuid, result));
    }

    public void startStep(final String parentUuid, final String uuid, final StepResult result) {
        notifier.beforeStepStart(result);
        result.setStage(Stage.RUNNING);
        result.setStart(System.currentTimeMillis());
        storage.startStep(uuid);
        storage.addStep(parentUuid, uuid, result);
        notifier.afterStepStart(result);
    }

    public void updateStep(final Consumer<StepResult> update) {
        storage.getCurrentStep().ifPresent(uuid -> updateStep(uuid, update));
    }

    public void updateStep(final String uuid, final Consumer<StepResult> update) {
        storage.getStep(uuid).ifPresent(step -> {
            notifier.beforeStepUpdate(step);
            update.accept(step);
            notifier.afterStepUpdate(step);
        });
    }

    public void stopStep() {
        storage.getCurrentStep().ifPresent(this::stopStep);
    }

    public void stopStep(final String uuid) {
        storage.removeStep(uuid).ifPresent(step -> {
            notifier.beforeStepStop(step);
            step.setStage(Stage.FINISHED);
            step.setStop(System.currentTimeMillis());
            storage.stopStep();
            notifier.afterStepStop(step);
        });
    }

    public void addAttachment(final String name, final String type,
                              final String fileExtension, final byte[] body) {
        addAttachment(name, type, fileExtension, new ByteArrayInputStream(body));
    }

    public void addAttachment(final String name, final String type,
                              final String fileExtension, final InputStream stream) {
        writeAttachment(prepareAttachment(name, type, fileExtension ,null), stream);
    }

    /**
     * 主要针对于自动截图功能( 会自动保存的 , 不用write)
     */
    public void addAttachment(String name ,String source){
        prepareAttachment(name ,"image/png" , ".png" ,source);
    }

    /**
     * 主要针对于删除文件后的处理操作 ( 删除关联的 Attachment )
     */
    public void removeAttachment(){
        Optional<String> currentStep = storage.getCurrentStep();
        currentStep.ifPresent(uuid ->
            storage.get(uuid ,WithAttachments.class).ifPresent(attachments ->
                    {
                        List<Attachment> att = attachments.getAttachments();
                        if (att.isEmpty()) {
                            Log.d(TAG,"empty Attachments ,can not remove it .");
                        }
                        att.remove(att.size() - 1);
                    }
            )
        );
    }

    @SuppressWarnings({"PMD.NullAssignment", "PMD.UseObjectForClearerAPI"})
    public String prepareAttachment(final String name, final String type, final String fileExtension , String source) {
        final Optional<String> currentStep = storage.getCurrentStep();
        currentStep.ifPresent(uuid -> Log.d(TAG,"Adding attachment to item with uuid "+uuid));
        final String extension = Optional.ofNullable(fileExtension)
                .filter(ext -> !ext.isEmpty())
                .map(ext -> ext.charAt(0) == '.' ? ext : "." + ext)
                .orElse("");
        if (isNone(source)) source = makeAttachmentSource(extension);
        final Attachment attachment = new Attachment()
                .withName(isEmpty(name) ? null : name)
                .withType(isEmpty(type) ? null : type)
                .withSource(source);

        currentStep.flatMap(uuid -> storage.get(uuid, WithAttachments.class))
                .ifPresent(withAttachments -> withAttachments.getAttachments().add(attachment));
        return attachment.getSource();
    }

    public void writeAttachment(final String attachmentSource, final InputStream stream) {
        writer.write(attachmentSource, stream);
    }

    public String makeAttachmentSource(String fileExtension){
        return UUID.randomUUID().toString() + ATTACHMENT_FILE_SUFFIX + fileExtension;
    }

    private boolean isEmpty(final String s) {
        return Objects.isNull(s) || s.isEmpty();
    }

}

File: allure2-android/src/main/java/com/yq/allure2_android/common/AllureLifecycle.kt
package com.yq.allure2_android.common

import android.util.Log
import java.io.ByteArrayInputStream
import java.io.InputStream
import java.util.*
import com.yq.allure2_android.android.listener.LifecycleNotifier
import com.yq.allure2_android.common.utils.AllureConstants.ATTACHMENT_FILE_SUFFIX
import com.yq.allure2_android.model.TestResult
import com.yq.allure2_android.common.resultRW.AllureResultsWriter
import com.yq.allure2_android.model.*


class AllureLifecycle(
        private val writer: AllureResultsWriter,
        private val storage: AllureStorage,
        private val notifier: LifecycleNotifier
) {
    val TAG = "allure_AllureLifecycle"

    fun startTestContainer(parentUuid: String, container: TestResultContainer) {
        updateTestContainer(parentUuid, { children.add(container.uuid) })
        startTestContainer(container)
    }

    fun startTestContainer(container: TestResultContainer) {
        notifier.beforeContainerStart(container)
        container.start = System.currentTimeMillis()
        storage.addContainer(container)
        notifier.afterContainerStart(container)
    }

    fun updateTestContainer(uuid: String, update: TestResultContainer.() -> Unit) {
        storage.getContainer(uuid)?.apply{
            notifier.beforeContainerUpdate(this)
            update()
            notifier.afterContainerUpdate(this)
        }
    }

    fun stopTestContainer(uuid: String) {
        storage.getContainer(uuid)?.apply {
            notifier.beforeContainerStop(this)
            this.stop = System.currentTimeMillis()
            notifier.afterContainerUpdate(this)
        }
    }

    fun writeTestContainer(uuid: String) {
        storage.removeContainer(uuid)?.apply{
            notifier.beforeContainerWrite(this)
            writer.write(this)
            notifier.afterContainerWrite(this)
        }
    }

    fun startPrepareFixture(parentUuid: String, uuid: String, result: FixtureResult) {
        notifier.beforeFixtureStart(result)
        updateTestContainer(parentUuid, { befores.add(result) })
        startFixture(uuid, result)
        notifier.afterFixtureStart(result)
    }

    fun startTearDownFixture(parentUuid: String, uuid: String, result: FixtureResult) {
        notifier.beforeFixtureStart(result)
        updateTestContainer(parentUuid, { afters.add(result) })
        startFixture(uuid, result)
        notifier.afterFixtureStart(result)
    }

    private fun startFixture(uuid: String, result: FixtureResult) {
        storage.addFixture(uuid, result)
        result.stage = Stage.RUNNING
        result.start = System.currentTimeMillis()
        storage.clearStepContext()
        storage.startStep(uuid)
    }

    fun updateFixture(update: FixtureResult.() -> Unit) {
        updateFixture(storage.getRootStep()!!, update)
    }

    fun updateFixture(uuid: String, update: FixtureResult.() -> Unit) {
        storage.getFixture(uuid).apply{
            notifier.beforeFixtureUpdate(this)
            update()
            notifier.afterFixtureUpdate(this)
        }
    }

    fun stopFixture(uuid: String) {
        storage.removeFixture(uuid).apply {
            notifier.beforeFixtureStop(this)
            storage.clearStepContext()
            this.stage = Stage.FINISHED
            this.stop = System.currentTimeMillis()
            notifier.afterFixtureStop(this)
        }
    }

    fun getCurrentTestCase(): String {
        return storage.getRootStep() ?: ""
    }

    fun scheduleTestCase(parentUuid: String, result: TestResult) {
        updateTestContainer(parentUuid, { children.add(result.uuid) })
        scheduleTestCase(result)
    }

    fun scheduleTestCase(result: TestResult) {
        notifier.beforeTestSchedule(result)
        result.stage = Stage.SCHEDULED
        storage.addTestResult(result)
        notifier.afterTestSchedule(result)
    }

    fun startTestCase(uuid: String) {
        storage.getTestResult(uuid)?.apply{ 
            notifier.beforeTestStart(this)
            this
                    .withStage(Stage.RUNNING)
                    .withStart(System.currentTimeMillis())
            storage.clearStepContext()
            storage.startStep(uuid)
            notifier.afterTestStart(this)
        }
    }

    fun updateTestCase(update: TestResult.()->Unit) {
        val uuid = storage.getRootStep()
        updateTestCase(uuid!!, update)
    }

    fun updateTestCase(uuid: String, update: TestResult.()->Unit) {
        storage.getTestResult(uuid)?.apply{
            notifier.beforeTestUpdate(this)
            update()
            notifier.afterTestUpdate(this)
        }
    }

    fun stopTestCase(uuid: String) {
        storage.getTestResult(uuid)?.apply{
            notifier.beforeTestStop(this)
            this
                    .withStage(Stage.FINISHED)
                    .withStop(System.currentTimeMillis())
            storage.clearStepContext()
            notifier.afterTestStop(this)
        }
    }

    fun writeTestCase(uuid: String) {
        storage.removeTestResult(uuid)?.apply{
            notifier.beforeTestWrite(this)
            writer.write(this)
            notifier.afterTestWrite(this)
        }
    }

    fun startStep(uuid: String, result: StepResult) {
        storage.getCurrentStep().apply{startStep(this, uuid, result) }
    }

    fun startStep(parentUuid: String, uuid: String, result: StepResult) {
        notifier.beforeStepStart(result)
        result.stage = Stage.RUNNING
        result.start = System.currentTimeMillis()
        storage.startStep(uuid)
        storage.addStep(parentUuid, uuid, result)
        notifier.afterStepStart(result)
    }

    fun updateStep(update: StepResult.()->Unit) {
        storage.getCurrentStep().apply{ updateStep(this, update) }
    }

    fun updateStep(uuid: String, update: StepResult.()->Unit) {
        storage.getStep(uuid).apply{
            notifier.beforeStepUpdate(this)
            update()
            notifier.afterStepUpdate(this)
        }
    }

    fun stopStep() {
        storage.getCurrentStep().let{ this.stopStep(it) }
    }

    fun stopStep(uuid: String) {
        storage.removeStep(uuid)?.apply{
            notifier.beforeStepStop(this)
            this.stage = Stage.FINISHED
            this.stop = System.currentTimeMillis()
            storage.stopStep()
            notifier.afterStepStop(this)
        }
    }

    fun addAttachment(name: String, type: String,
                      fileExtension: String, body: ByteArray) {
        addAttachment(name, type, fileExtension, ByteArrayInputStream(body))
    }

    fun addAttachment(name: String, type: String?,fileExtension: String?, stream: InputStream) {
        writeAttachment(prepareAttachment(name, type, fileExtension), stream)
    }

    /**
     * 主要针对于自动截图功能( 会自动保存的 , 不用write)
     */
    fun addAttachment(name: String ,source: String?){
        prepareAttachment(name ,"image/png" , ".png" ,source)
    }

    /**
     * 主要针对于删除文件后的处理操作 ( 删除关联的 Attachment )
     */
    fun removeAttachment(){
        val currentStep = storage.getCurrentStep()
        storage.get(currentStep, WithAttachments::class.java).attachments?.apply {
            if (this.isEmpty()){
                println("empty Attachments ,can not remove it .")
                return
            }
            this.removeAt(this.size - 1)
        }
    }
    /**
     * @param name 指定的名称 ( My attachment )
     * @param type 指定的类型 ( text/plain )
     * @param fileExtension 扩展名 ( .txt )
     */
    fun prepareAttachment(name: String, type: String?, fileExtension: String?, source: String? = null): String {
        val currentStep = storage.getCurrentStep()
        currentStep.let{ Log.d(TAG,"Adding attachment to item with uuid $it") }
        val attachment = Attachment()
                .withName(if (isEmpty(name)) null else name)
                .withType(if (isEmpty(type)) null else type)
                .withSource(source?: makeAttachmentSource(fileExtension))
        storage.get(currentStep, WithAttachments::class.java).attachments.add(attachment)
        return attachment.source
    }

    fun writeAttachment(attachmentSource: String, stream: InputStream) {
        writer.write(attachmentSource, stream)
    }

    fun makeAttachmentSource(fileExtension: String?):String{
        return UUID.randomUUID().toString() + ATTACHMENT_FILE_SUFFIX + fileExtension
    }
    private fun isEmpty(s: String?): Boolean {
        return s==null || s.isEmpty()
    }


}